src/phoenix/server/api/routers/trace_handler.py

```diff
@@ -54,7 +60,122 @@ async def post(self, request: Request) -> Response:
         for resource_spans in req.resource_spans:
             project_name = get_project_name(resource_spans.resource.attributes)
             for scope_span in resource_spans.scope_spans:
-                for span in scope_span.spans:
-                    self.traces.put(decode(span), project_name=project_name)
+                for otlp_span in scope_span.spans:
+                    span = decode(otlp_span)
+                    async with self.db() as session:
+                        await _insert_span(session, span, project_name)
+                    self.traces.put(span, project_name=project_name)
                     await asyncio.sleep(0)
         return Response()
+
+
+async def _insert_span(session: AsyncSession, span: Span, project_name: str) -> None:
+    if not (
+        project_rowid := await session.scalar(
+            text("SELECT rowid FROM projects WHERE name = :name;"),
+            {"name": project_name},
+        )
+    ):
+        project_rowid = await session.scalar(
+            text("INSERT INTO projects(name) VALUES(:name) RETURNING rowid;"),
+            {"name": project_name},
+        )
```

Comment:
this makes it so that you have to hit the database on every insert. I think this is where we can build a bit of optimization via dynamic programming where you can keep a list of seen projects. If you've seen it before you don't need to insert.

Reply:
yea, that makes sense. we may have to cache it outside the handler though